src/config/format.rs

```diff
@@ -0,0 +1,258 @@
+//! Support for loading configs from multiple formats.
+
+#![deny(missing_docs, missing_debug_implementations)]
+
+use serde::de;
+use std::path::Path;
+
+/// The format used to represent the configuration data.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+pub enum Format {
+    /// The format could not be determined.
+    Unknown,
+    /// TOML format is used.
+    TOML,
+    /// JSON format is used.
+    JSON,
+    /// YAML format is used.
+    YAML,
+}
+
+impl Default for Format {
+    fn default() -> Self {
+        Format::TOML
+    }
+}
+
+impl Format {
+    fn refute_unknown(self) -> Self {
+        match self {
+            Format::Unknown => Format::default(),
+            _ => self,
+        }
+    }
+}
+
+/// Obtain the format from the file path using extension as a hint.
+pub fn from_path(path: impl AsRef<Path>) -> Format {
+    match path.as_ref().extension().and_then(|ext| ext.to_str()) {
+        Some("toml") => Format::TOML,
+        Some("yaml") | Some("yml") => Format::YAML,
+        Some("json") => Format::JSON,
+        _ => Format::Unknown,
+    }
+}
+
+/// Parse the string represented in the specified format.
+/// If the format is unknown - fallback to the default format and attempt
+/// parsing using that.
+pub fn deserialize<T>(content: &str, format: Format) -> Result<T, Vec<String>>
+where
+    T: de::DeserializeOwned,
+{
+    match format.refute_unknown() {
+        Format::Unknown => unreachable!(),
+        Format::TOML => toml::from_str(content).map_err(|e| vec![e.to_string()]),
+        Format::YAML => serde_yaml::from_str(content).map_err(|e| vec![e.to_string()]),
+        Format::JSON => serde_json::from_str(content).map_err(|e| vec![e.to_string()]),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::config::ConfigBuilder;
+
+    use super::*;
+
+    /// This test ensures the logic to guess file format from the file path
+    /// works correctly.
+    /// Like all other tests, it also demonstrates various cases and how our
+    /// code behaves when it enounters them.
+    #[test]
+    fn test_from_path() {
+        let cases = vec![
+            // Unknown - odd variants.
+            ("", Format::Unknown),
+            (".", Format::Unknown),
+            // Unknown - no ext.
+            ("myfile", Format::Unknown),
+            ("mydir/myfile", Format::Unknown),
+            ("/mydir/myfile", Format::Unknown),
+            // Unknown - some unknown ext.
+            ("myfile.myext", Format::Unknown),
+            ("mydir/myfile.myext", Format::Unknown),
```

Comment:
The default is TOML no? Why do we need these tests for this format unknown?

Reply:
These tests are for the function that determines the format from the file extension.

There can be other functions like this, and the flow would look something like this:
1. When we have the file name (the relevant context), attempt to detect the format for the first time and remember it.
2. When we have the file content (different context), if the file format is still unknown - attempt to detect the file format again from the content, and preserve the format.
3. When we get to actually using the format - fallback to TOML if still unknown, and apply deserialization for the requested format.

Reply:
Since we only support toml, yaml, or json, perhaps attempting to open a file with a different extension should result in an error for now. If we add support for new file types that can remove the error for that type.